[Infra UI] Convert bytes to bits before formatting for bits

simianhacker opened the pull request:
## Summary

This PR attempts to fix #40138 by converting bytes to bits when the formatter is using a bit format. This also attempts to make clear that the number being formatted is expected to be bytes by changing the `val` argument to `bytes`.

### Checklist

Use ~~strikethroughs~~ to remove checklist items you don't feel are applicable to this PR.

- ~~This was checked for cross-browser compatibility, [including a check against IE11](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#cross-browser-compatibility)~~
- ~~Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/master/packages/kbn-i18n/README.md)~~
- ~~[Documentation](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#writing-documentation) was added for features that require explanation or tutorials~~
- [x] [Unit or functional tests](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#cross-browser-compatibility) were updated or added to match the most common scenarios
- ~~This was checked for [keyboard-only and screenreader accessibility](https://developer.mozilla.org/en-US/docs/Learn/Tools_and_testing/Cross_browser_testing/Accessibility#Accessibility_testing_checklist)~~

skh commented:
Nice catch!
***
I have two questions. The first on is not related to this change. In `public/lib/lib.ts` we have these different types of output data defined:

```
export enum InfraWaffleMapDataFormat {
  bytesDecimal = 'bytesDecimal',
  bytesBinaryIEC = 'bytesBinaryIEC',
  bytesBinaryJEDEC = 'bytesBinaryJEDEC',
  bitsDecimal = 'bitsDecimal',
  bitsBinaryIEC = 'bitsBinaryIEC',
  bitsBinaryJEDEC = 'bitsBinaryJEDEC',
  abbreviatedNumber = 'abbreviatedNumber',
}
```
Of these, `public/utils/formatters/index.ts` only uses
```
InfraWaffleMapDataFormat.abbreviatedNumber
InfraWaffleMapDataFormat.bytesDecimal
InfraWaffleMapDataFormat.bitsDecimal
```
The others are only used in the tests, as far as I could find out. If this is indeed the case, why do we need them? Removing them would make the data formatter less powerful, but also less intimidating to read.
***
My second question is more a concern: the data formatter implicitly assumes that it is given `bytes` to format -- it is now less implicit as you renamed the input variable to `bytes`, but it is still only clear when one looks at the implementation. Then, if the output format is any of the `bit` formats, it now silently converts bits to bytes to fix the issue that triggered this change.

The way the formatters are created in `public/utils/formatters/index.ts ` gives me the impression, however, that one is for bits and one is for bytes:
```
  [InfraFormatterType.bytes]: createDataFormatter(InfraWaffleMapDataFormat.bytesDecimal),
  [InfraFormatterType.bits]: createDataFormatter(InfraWaffleMapDataFormat.bitsDecimal),
```
As a user of this formatter in other parts of the code I would probably have converted `bytes` to `bits` myself first, and then used the formatter, assuming that it only formats according to the slightly different rules for `bits` / `bytes` and otherwise leaves the data alone.

In short, it is making me nervous that data is converted in a formatter, and I would prefer having the `bytes` to `bits` conversion somewhere more explicitly done.
***
I also checked whether there could be data in `bits` coming in from `metricbeat` with
```
skh@noether:~/projects/beats/metricbeat$ grep -r "format: " *
```
(where `~/project/beats` is a checkout of `master` of https://github.com/elastic/beats/). This catches all field definitions in the various `_meta/fields.yml` files in `metricbeat` that have a data type, for example in `metricbeat/module/system/network/_meta/fields.yml` the definition of `system.network.out.bytes`:
```
    - name: out.bytes
      type: long
      format: bytes
      description: >
        The number of bytes sent.
```
The result of my query above is that, indeed, we have many fields in `bytes` format, and none in `bits`, so for now the behaviour of the UI is probably correct, as long as it only deals with `metricbeat` data. But even then, I still think the conversion shouldn't happen silently in the formatter. Also, we might one day be confronted with input in `bits` from other sources.
***
An alternative would be to make it clear in the naming that this formatter expects `bytes` as input, e.g. by renaming the file it resides in from `data.ts` to `bytes.ts` and adding a bit of commentary around it.

skh commented on x-pack/legacy/plugins/infra/public/utils/formatters/data.ts:
```diff
@@ -62,12 +62,19 @@ const BASES = {
   const base = BASES[format];
-  const power = Math.min(Math.floor(Math.log(Math.abs(val)) / Math.log(base)), labels.length - 1);
+  const value = BIT_BASES.includes(format) ? bytes * 8 : bytes;
+  const power = Math.min(Math.floor(Math.log(Math.abs(value)) / Math.log(base)), labels.length - 1);
```
What does this calculation do, and why? Could you add a comment?

weltenwort replied:
As I read it, it determines the `power` in `base**power` given a `base` to decide which abbreviated unit to use (kilo, mega, ..., kibi, mebi, ...) for a `value`.

simianhacker replied:
I added a comment. I'm using an [exponential equation](http://www.montereyinstitute.org/courses/DevelopmentalMath/COURSE_TEXT2_RESOURCE/U18_L4_T1_text_final.html) to figure out the power used for the label. If the power is greater then the max label then we just use the max label.

simianhacker commented:
> The others are only used in the tests, as far as I could find out. If this is indeed the case, why do we need them? Removing them would make the data formatter less powerful, but also less intimidating to read.

@skh Originally, I was only going to implement `bytesDecimal` and `bitsDecimal` but at the time someone on the team (I forget who) mentioned that technically there are more formats and that at some point in the future we might want to enable the user to configure which format they wanted specifically. So in response I added all the formats I could find with regards to formatting data. I agree this is over engineered.

> My second question is more a concern: the data formatter implicitly assumes that it is given bytes to format -- it is now less implicit as you renamed the input variable to bytes, but it is still only clear when one looks at the implementation. Then, if the output format is any of the bit formats, it now silently converts bytes to bits to fix the issue that triggered this change.

@skh Since beats ships all data in bytes, I think we should make it clear that our expectation is `bytes -> format`. I agree we should we changed the `createDataFormatter` method to `createBytesFormatter` and then change the file name to `bytes.ts`.  I can also add some comments to make it clear that we are always expecting bytes and that if you're using the `bit` formatter it's going to convert from bytes to bits then format. If we set the expectation of bytes in format out, then the fact that it's converting to bits internally is just an implementation detail.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/6373/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/6399/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/6431/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/6463/)

elasticmachine commented:
Pinging @elastic/infra-logs-ui

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/6899/)
